pbft/pbft_checkpoint_manager.cpp

```diff
@@ -0,0 +1,323 @@
+        {
+            msg_pair.second.destroy();
+        }
+    }
+
+    // now remove them from memory as well
+    for (auto it = this->partial_checkpoint_proofs.begin(); it != this->partial_checkpoint_proofs.end();)
+    {
+        auto seq = it->first.first;
+        if (seq > cp.first)
+        {
+            it++;
+        }
+        else
+        {
+            it = this->partial_checkpoint_proofs.erase(it);
+        }
+    }
+}
+
+void
+pbft_checkpoint_manager::send_delayed_state_request(const checkpoint_t& cp)
+{
+    this->trigger_catchup_timer = this->io_context->make_unique_steady_timer();
+    this->trigger_catchup_timer->expires_from_now(CHECKPOINT_CATCHUP_GRACE_PERIOD);
+    this->trigger_catchup_timer->async_wait(
+            [cp, weak_this = weak_from_this()](auto ec)
+            {
+                if (ec)
+                {
+                    return;
+                }
+
+                auto strong_this = weak_this.lock();
+
+                if (strong_this && strong_this->get_latest_stable_checkpoint() == cp)
+                {
+                    std::lock_guard<std::mutex> lock(strong_this->lock);
+                    strong_this->send_state_request();
+                }
+            }
+    );
+}
+
+void
+pbft_checkpoint_manager::send_state_request()
+{
+    pbft_membership_msg msg;
+    msg.set_type(PBFT_MMSG_GET_STATE);
+    msg.set_sequence(this->latest_stable_checkpoint.value().first);
+    msg.set_state_hash(this->latest_stable_checkpoint.value().second);
+
+    uint32_t selected = pbft::generate_random_number(0, this->latest_stable_checkpoint_proof.size() - 1);
+    auto it = this->latest_stable_checkpoint_proof.begin();
+    std::advance(it, selected);
+    auto selected_peer = (*it).first;
+
+    LOG(info) << "Requesting state at " << msg.sequence() << " from " << selected_peer;
+
+    auto msg_ptr = std::make_shared<bzn_envelope>();
+    msg_ptr->set_pbft_membership(msg.SerializeAsString());
+
+    this->node->send_signed_message(selected_peer, msg_ptr);
+}
+
+void
+pbft_checkpoint_manager::init_persists()
+{
+    persistent<std::string>::init_kv_container<bzn::uuid_t>(this->storage, STABLE_CHECKPOINT_PROOF_KEY,
+            this->latest_stable_checkpoint_proof);
+    persistent<std::string>::init_kv_container2<bzn::uuid_t, checkpoint_t>(this->storage, PARTIAL_CHECKPOINT_PROOFS_KEY,
+            this->partial_checkpoint_proofs);
+}
+
+std::unordered_map<bzn::uuid_t, std::string>
+pbft_checkpoint_manager::get_latest_stable_checkpoint_proof() const
+{
+    std::lock_guard<std::mutex> lock(this->lock);
+    std::unordered_map<bzn::uuid_t, std::string> result;
+    for (const auto& pair : this->latest_stable_checkpoint_proof)
+    {
+        result[pair.first] = pair.second.value();
+    }
+
+    return result;
+}
+
+size_t
+pbft_checkpoint_manager::partial_checkpoint_proofs_count() const
+{
+    std::lock_guard<std::mutex> lock(this->lock);
+    return this->partial_checkpoint_proofs.size();
+
+}
```

Comment:
compilers don't dig modules with no empty space at the back. Meh, fix it later...